examples/chat2-reliable/reliability.go

```diff
@@ -0,0 +1,458 @@
+	filter   *bloom.BloomFilter
+	window   time.Duration
+	messages []TimestampedMessageID
+	mutex    sync.Mutex
+}
+
+func NewRollingBloomFilter() *RollingBloomFilter {
+	return &RollingBloomFilter{
+		filter:   bloom.NewWithEstimates(bloomFilterSize, bloomFilterFPRate),
+		window:   bloomFilterWindow,
+		messages: make([]TimestampedMessageID, 0),
+	}
+}
+
+func (rbf *RollingBloomFilter) Add(messageID string) {
+	rbf.mutex.Lock()
+	defer rbf.mutex.Unlock()
+
+	rbf.filter.Add([]byte(messageID))
+	rbf.messages = append(rbf.messages, TimestampedMessageID{
+		ID:        messageID,
+		Timestamp: time.Now(),
+	})
+}
+
+func (rbf *RollingBloomFilter) Test(messageID string) bool {
+	rbf.mutex.Lock()
+	defer rbf.mutex.Unlock()
+
+	return rbf.filter.Test([]byte(messageID))
+}
+
+func (rbf *RollingBloomFilter) Clean() {
+	rbf.mutex.Lock()
+	defer rbf.mutex.Unlock()
+
+	cutoff := time.Now().Add(-rbf.window)
+	newMessages := make([]TimestampedMessageID, 0)
+	newFilter := bloom.NewWithEstimates(bloomFilterSize, bloomFilterFPRate)
+
+	for _, msg := range rbf.messages {
+		if msg.Timestamp.After(cutoff) {
+			newMessages = append(newMessages, msg)
+			newFilter.Add([]byte(msg.ID))
+		}
+	}
+
+	rbf.messages = newMessages
+	rbf.filter = newFilter
+}
+
+func (c *Chat) startBloomFilterCleaner() {
+	defer c.wg.Done()
+
+	ticker := time.NewTicker(bloomFilterCleanInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-c.ctx.Done():
+			return
+		case <-ticker.C:
+			c.bloomFilter.Clean()
+		}
+	}
+}
+
+func (c *Chat) processReceivedMessage(msg *pb.Message) {
+	// Check if the message is already in the bloom filter
+	if c.bloomFilter.Test(msg.MessageId) {
+		// Review ACK status of messages in the unacknowledged outgoing buffer
+		c.reviewAckStatus(msg)
+		return
+	}
+
+	// Update Lamport timestamp
+	c.updateLamportTimestamp(msg.LamportTimestamp)
+
+	// Update bloom filter
+	c.bloomFilter.Add(msg.MessageId)
+
+	// Check causal dependencies
+	missingDeps := c.checkCausalDependencies(msg)
+	if len(missingDeps) == 0 {
+		if msg.Content != "" {
+			// Process the message
+			c.ui.ChatMessage(int64(c.getLamportTimestamp()), msg.SenderId, msg.Content)
+			// Add to message history
+			c.addToMessageHistory(msg)
+		}
+
+		// Process any messages in the buffer that now have their dependencies met
+		c.processBufferedMessages()
+	} else {
+		// Request missing dependencies
+		for _, depID := range missingDeps {
+			c.requestMissingMessage(depID)
+		}
+		// Add to incoming buffer
+		c.addToIncomingBuffer(msg)
+	}
+}
+
+func (c *Chat) processBufferedMessages() {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	processed := make(map[string]bool)
+	remainingBuffer := make([]*pb.Message, 0, len(c.incomingBuffer))
+
+	for _, msg := range c.incomingBuffer {
+		if processed[msg.MessageId] {
+			continue
+		}
+
+		missingDeps := c.checkCausalDependencies(msg)
+		if len(missingDeps) == 0 {
+			// Release the lock while processing the message
+			c.mutex.Unlock()
```

Comment:
Hmm, I feel like this might mess up with the for loop in case the `incommingBuffer` gets changed while unlocked. I guess we need separate locs for these crutial attributes